From 1d66b11f1f0a08acfa0fc10660bac57b29d3da25 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Matti=20Lehtim=C3=A4ki?= <matti.lehtimaki@jolla.com>
Date: Wed, 11 Oct 2023 15:43:59 +0000
Subject: [PATCH 1/3] (hybris) Prevent using dm-verity for apexes

Patch apexd.rc to use apexd from /usr/libexec/droid-hybris/system/bin/.
Disable the forced reboot to make shutdown work.

Change-Id: Ie537eb6458022abf581d66dfc6f1445029cbedd9
---
 apexd/apexd.cpp     |  3 ++-
 apexd/apexd.rc      |  1 +
 apexd/apexd_utils.h | 10 +++++-----
 3 files changed, 8 insertions(+), 6 deletions(-)

diff --git a/apexd/apexd.cpp b/apexd/apexd.cpp
index 8e539cfd..3d899f31 100644
--- a/apexd/apexd.cpp
+++ b/apexd/apexd.cpp
@@ -537,13 +537,14 @@ Result<MountedApexData> MountPackageImpl(const ApexFile& apex,
   // dm-verity because they are already in the dm-verity protected partition;
   // system. However, note that we don't skip verification to ensure that APEXes
   // are correctly signed.
-  const bool mount_on_verity = !instance.IsPreInstalledApex(apex) ||
+  bool mount_on_verity = !instance.IsPreInstalledApex(apex) ||
                                // decompressed apexes are on /data
                                instance.IsDecompressedApex(apex) ||
                                // block apexes are from host
                                instance.IsBlockApex(apex);
 
   DmVerityDevice verity_dev;
+  mount_on_verity = false;
   loop::LoopbackDeviceUniqueFd loop_for_hash;
   if (mount_on_verity) {
     std::string hash_device = loopback_device.name;
diff --git a/apexd/apexd.rc b/apexd/apexd.rc
index af07a1b6..f68bb533 100644
--- a/apexd/apexd.rc
+++ b/apexd/apexd.rc
@@ -5,6 +5,7 @@ service apexd /system/bin/apexd
     group system
     oneshot
     disabled # does not start with the core class
+    # Disabled for hybris shutdown to work correctly
     reboot_on_failure reboot,apexd-failed
     # CAP_CHOWN, CAP_DAC_OVERRIDE, CAP_DAC_READ_SEARCH required for apexdata snapshot & restore
     # CAP_SYS_ADMIN is required to access device-mapper and to use mount syscall
diff --git a/apexd/apexd_utils.h b/apexd/apexd_utils.h
index d2d6566e..8c8c9a4d 100644
--- a/apexd/apexd_utils.h
+++ b/apexd/apexd_utils.h
@@ -306,11 +306,11 @@ inline android::base::Result<uintmax_t> GetFileSize(
   return value;
 }
 
-inline android::base::Result<void> RestoreconPath(const std::string& path) {
-  unsigned int seflags = SELINUX_ANDROID_RESTORECON_RECURSE;
-  if (selinux_android_restorecon(path.c_str(), seflags) < 0) {
-    return android::base::ErrnoError() << "Failed to restorecon " << path;
-  }
+inline android::base::Result<void> RestoreconPath(const std::string&) {
+  //unsigned int seflags = SELINUX_ANDROID_RESTORECON_RECURSE;
+  //if (selinux_android_restorecon(path.c_str(), seflags) < 0) {
+  //  return android::base::ErrnoError() << "Failed to restorecon " << path;
+  //}
   return {};
 }
 
-- 
2.55.0

